tidy3d/components/heat_charge/source.py

```diff
@@ -0,0 +1,94 @@
+"""Defines heat-charge material specifications for 'HeatChargeSimulation'"""
+from __future__ import annotations
+
+from abc import ABC
+from typing import Union, Tuple
+
+import pydantic.v1 as pd
+
+from .viz import plot_params_heat_source
+
+from ..base import cached_property
+from ..base_sim.source import AbstractSource
+from ..data.data_array import TimeDataArray
+from ..viz import PlotParams
+
+from ...constants import VOLUMETRIC_HEAT_RATE
+from ...exceptions import SetupError
+from ...log import log
+
+
+class HeatChargeSource(AbstractSource, ABC):
+    """Abstract source for heat-charge simulations. All source types
+    for 'HeatChargeSimulation' derive from this class."""
+
+    structures: Tuple[str, ...] = pd.Field(
+        title="Target Structures",
+        description="Names of structures where to apply heat source.",
+    )
+
+    @cached_property
+    def plot_params(self) -> PlotParams:
+        """Default parameters for plotting a Source object."""
+        return plot_params_heat_source
+
+    @pd.validator("structures", always=True)
+    def check_non_empty_structures(cls, val):
+        """Error if source doesn't point at any structures."""
+        if len(val) == 0:
+            raise SetupError("List of structures for heat source is empty.")
+
+        return val
+
+
+class HeatSource(HeatChargeSource):
+    """Adds a volumetric heat source (heat sink if negative values
+    are provided) to specific structures in the scene.
+
+    Example
+    -------
+    >>> heat_source = HeatSource(rate=1, structures=["box"])
+    """
+
+    rate: Union[float, TimeDataArray] = pd.Field(
+        title="Volumetric Heat Rate",
+        description="Volumetric rate of heating or cooling (if negative) in units of "
+        f"{VOLUMETRIC_HEAT_RATE}.",
+        units=VOLUMETRIC_HEAT_RATE,
+    )
+
+
+class HeatFromElectricSource(HeatChargeSource):
+    """Volumetric heat source generated from an electric simulation.
+    If a `HeatFromElectricSource` is specified as a source, appropriate boundary
+    conditions for an electric simulation must be provided, since such a simulation
+    will be executed before the heat simulation can run.
+
+    Example
+    -------
+    >>> heat_source = HeatFromElectricSource(structures=["box"])
```

Comment:
so, do we still need to provide a list of structures for `HeatFromElectricSource`?

Reply:
On this line, and given what we've chatted about in the slack, does `HeatFromChargeSource` mantain naming consistency with the `charge` based approach?

Reply:
This branch still requires `structures`. Let me update this behavior from `marc/electrostatic-dev`